cluster-autoscaler/debuggingsnapshot/debugging_snapshotter.go

```diff
@@ -0,0 +1,238 @@
+	// it states that a snapshot request needs to be processed
+	TRIGGER_ENABLED
+	// START_DATA_COLLECTION is used to synchronise the collection of data.
+	// Since the trigger is an asynchronous process, data collection could be started mid-loop
+	// leading to incomplete data. So setter methods wait for START_DATA_COLLECTION before collecting data
+	// which is set at the start of the next loop after receiving the trigger
+	START_DATA_COLLECTION
+	// DATA_COLLECTED is set by setter func (also used by setter func for data collection)
+	// This is set to let Flush know that at least some data collected and there isn't
+	// an error State leading to no data collection
+	DATA_COLLECTED
+)
+
+// DebuggingSnapshotterImpl is the impl for DebuggingSnapshotter
+type DebuggingSnapshotterImpl struct {
+	// State captures the internal state of the snapshotter
+	State *DebuggingSnapshotterState
+	// DebuggingSnapshot is the data bean for the snapshot
+	DebuggingSnapshot DebuggingSnapshot
+	// Mutex is the synchronisation used to the methods/states in the critical section
+	Mutex *sync.Mutex
+	// Trigger is the channel on which the Response Handler waits on to know
+	// when there is data to be flushed back to the channel from the Snapshot
+	Trigger chan struct{}
+	// CancelRequest is the cancel function for the snapshot request. It is used to
+	// terminate any ongoing request when CA is shutting down
+	CancelRequest context.CancelFunc
+}
+
+// DebuggingSnapshotter is the interface for debugging snapshot
+type DebuggingSnapshotter interface {
+
+	// StartDataCollection will check the State(s) and enable data
+	// collection for the loop if applicable
+	StartDataCollection()
+	// SetNodeGroupInfo is a setter to capture all the NodeInfo
+	SetNodeGroupInfo([]*framework.NodeInfo)
+	// SetUnscheduledPodsCanBeScheduled is a setter for all pods which are unscheduled
+	// but they can be scheduled. i.e. pods which aren't triggering scale-up
+	SetUnscheduledPodsCanBeScheduled([]*v1.Pod)
+	// ResponseHandler is the http response handler to manage incoming requests
+	ResponseHandler(http.ResponseWriter, *http.Request)
+	// IsDataCollectionAllowed checks the internal State of the snapshotter
+	// to find if data can be collected. This can be used before preprocessing
+	// for the snapshot
+	IsDataCollectionAllowed() bool
+	// Flush triggers the flushing of the snapshot
+	Flush()
+	// Cleanup clears the internal data beans of the snapshot, readying for next request
+	Cleanup()
+}
+
+// NewDebuggingSnapshotter returns a new instance of DebuggingSnapshotter
+func NewDebuggingSnapshotter(isDebuggerEnabled bool) DebuggingSnapshotter {
+	state := SNAPSHOTTER_DISABLED
+	if isDebuggerEnabled {
+		klog.Infof("Debugging Snapshot is enabled")
+		state = LISTENING
+	}
+	return &DebuggingSnapshotterImpl{
+		State:             &state,
+		Mutex:             &sync.Mutex{},
+		DebuggingSnapshot: &DebuggingSnapshotImpl{},
+		Trigger:           make(chan struct{}, 1),
```

Comment:
I don't think this needs to be buffered anymore.

Reply:
This needs to be buffered or else putting on the channel becomes a blocking call (instead of non-blocking) on when the channel is emptied. [this was evident in the test]